File: myExpenses/src/main/java/org/totschnig/myexpenses/adapter/SyncBackendAdapter.java
package org.totschnig.myexpenses.adapter;

import android.content.Context;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.ExpandableListView;
import android.widget.ImageView;
import android.widget.TextView;

import com.annimon.stream.Exceptional;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.model.Account;
import org.totschnig.myexpenses.model.CurrencyContext;
import org.totschnig.myexpenses.sync.json.AccountMetaData;

import java.util.List;
import java.util.Map;

import androidx.annotation.Nullable;
import androidx.core.util.Pair;

public class SyncBackendAdapter extends BaseExpandableListAdapter {

  public enum SyncState {
    SYNCED_TO_THIS,
    SYNCED_TO_OTHER,
    UNSYNCED,
    UNKNOWN,
    ERROR
  }

  private List<Pair<String, Boolean>> syncAccounts;
  private final SparseArray<List<Exceptional<AccountMetaData>>> accountMetaDataMap = new SparseArray<>();
  private final LayoutInflater layoutInflater;
  private Map<String, String> localAccountInfo;
  private final CurrencyContext currencyContext;

  public SyncBackendAdapter(Context context, CurrencyContext currencyContext, List<Pair<String, Boolean>> syncAccounts) {
    this.layoutInflater = LayoutInflater.from(context);
    this.syncAccounts = syncAccounts;
    this.currencyContext = currencyContext;
  }

  @Override
  public Object getChild(int groupPosition, int childPosititon) {
    List<Exceptional<AccountMetaData>> childList = getChildList(groupPosition);
    return childList != null ? childList.get(childPosititon) : null;
  }

  private List<Exceptional<AccountMetaData>> getChildList(int groupPosition) {
    return accountMetaDataMap.get(groupPosition);
  }

  @Override
  public long getChildId(int groupPosition, int childPosition) {
    return childPosition;
  }

  @Override
  public View getChildView(int groupPosition, final int childPosition,
                           boolean isLastChild, View convertView, ViewGroup parent) {
    Exceptional<AccountMetaData> accountMetaData = (Exceptional<AccountMetaData>) getChild(groupPosition, childPosition);

    if (convertView == null) {
      convertView = layoutInflater.inflate(R.layout.sync_account_row, parent, false);
    }
    ImageView syncStateView = convertView.findViewById(R.id.state);
    final TextView labelTextView = convertView.findViewById(R.id.label);
    if (accountMetaData.isPresent()) {
      labelTextView.setText(accountMetaData.get().toString());
      convertView.findViewById(R.id.color1).setBackgroundColor(accountMetaData.get().color());
      SyncState syncState = getSyncState(groupPosition, childPosition);
      switch (syncState) {
        case UNKNOWN:
          syncStateView.setVisibility(View.GONE);
          break;
        case SYNCED_TO_THIS:
          syncStateView.setVisibility(View.VISIBLE);
          syncStateView.setImageResource(R.drawable.ic_sync);
          break;
        case UNSYNCED:
        case SYNCED_TO_OTHER:
          syncStateView.setVisibility(View.VISIBLE);
          syncStateView.setImageResource(R.drawable.ic_action_sync_unlink);
          break;
      }
    } else {
      syncStateView.setVisibility(View.GONE);
      labelTextView.setText(accountMetaData.getException().getMessage());
    }
    return convertView;
  }

  @Override
  public int getChildrenCount(int groupPosition) {
    List<Exceptional<AccountMetaData>> childList = getChildList(groupPosition);
    return childList != null ? childList.size() : 0;
  }

  @Override
  public Object getGroup(int groupPosition) {
    return syncAccounts.get(groupPosition);
  }

  public String getBackendLabel(int groupPosition) {
    return ((Pair<String, Boolean>) getGroup(groupPosition)).first;
  }

  @Override
  public int getGroupCount() {
    return syncAccounts.size();
  }

  @Override
  public long getGroupId(int groupPosition) {
    return groupPosition;
  }

  @Override
  public View getGroupView(int groupPosition, boolean isExpanded,
                           View convertView, ViewGroup parent) {
    final Pair<String, Boolean> group = (Pair<String, Boolean>) getGroup(groupPosition);
    String headerTitle = group.first;
    if (convertView == null) {
      convertView = layoutInflater.inflate(R.layout.sync_backend_row, parent, false);
    }

    ((TextView) convertView.findViewById(R.id.label)).setText(headerTitle);
    convertView.findViewById(R.id.state).setVisibility(group.second ? View.VISIBLE : View.GONE);

    return convertView;
  }

  @Override
  public boolean hasStableIds() {
    return false;
  }

  @Override
  public boolean isChildSelectable(int groupPosition, int childPosition) {
    return true;
  }

  public void setAccountList(List<Pair<String, Boolean>> accountList) {
    syncAccounts = accountList;
    accountMetaDataMap.clear();
    notifyDataSetChanged();
  }

  public void setAccountMetadata(int groupPosition, List<Exceptional<AccountMetaData>> accountMetaDataList) {
    accountMetaDataMap.put(groupPosition, accountMetaDataList);
    notifyDataSetChanged();
  }

  public boolean hasAccountMetadata(int groupPosition) {
    return accountMetaDataMap.get(groupPosition) != null;
  }

  public void setLocalAccountInfo(Map<String, String> uuid2syncMap) {
    localAccountInfo = uuid2syncMap;
    notifyDataSetChanged();
  }

  public SyncState getSyncState(long packedPosition) {
    return getSyncState(ExpandableListView.getPackedPositionGroup(packedPosition), ExpandableListView.getPackedPositionChild(packedPosition));
  }

  private SyncState getSyncState(int groupPosition, int childPosition) {
    String syncAccount = getBackendLabel(groupPosition);
    Exceptional<AccountMetaData> accountMetaData = (Exceptional<AccountMetaData>) getChild(groupPosition, childPosition);
    if (accountMetaData.isPresent()) {
      String uuid = accountMetaData.get().uuid();
      if (localAccountInfo != null && localAccountInfo.containsKey(uuid)) {
        if (localAccountInfo.get(uuid) == null) {
          return SyncState.UNSYNCED;
        }
        return syncAccount.equals(localAccountInfo.get(uuid)) ?
            SyncState.SYNCED_TO_THIS : SyncState.SYNCED_TO_OTHER;
      }
      return SyncState.UNKNOWN;
    } else {
      return SyncState.ERROR;
    }
  }

  @Nullable
  public Account getAccountForSync(long packedPosition) {
    int groupPosition = ExpandableListView.getPackedPositionGroup(packedPosition);
    final Exceptional<AccountMetaData> child = (Exceptional<AccountMetaData>) getChild(groupPosition,
        ExpandableListView.getPackedPositionChild(packedPosition));
    if (child.isPresent()) {
      Account account = child.get().toAccount(currencyContext);
      account.setSyncAccountName(getBackendLabel(groupPosition));
      return account;
    } else {
      return null;
    }
  }

  public String getSyncAccountName(long packedPosition) {
    int groupPosition = ExpandableListView.getPackedPositionGroup(packedPosition);
    return getBackendLabel(groupPosition);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/adapter/SyncBackendAdapter.kt
package org.totschnig.myexpenses.adapter

import android.content.Context
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.ExpandableListView
import android.widget.ImageView
import android.widget.TextView
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.model2.Account
import org.totschnig.myexpenses.sync.json.AccountMetaData

data class LocalAccountInfo(val uuid: String, val syncAccountName: String?, val isSealed: Boolean)

class SyncBackendAdapter(
    context: Context,
    private val homeCurrency: String,
    private var syncAccounts: List<Pair<String, Boolean>>
) : BaseExpandableListAdapter() {
    enum class SyncState {
        SYNCED_TO_THIS, SYNCED_TO_OTHER, UNSYNCED, SEALED, UNKNOWN, ERROR
    }

    private val accountMetaDataMap = SparseArray<List<Result<AccountMetaData>>?>()
    private val layoutInflater = LayoutInflater.from(context)
    private var localAccountInfo: List<LocalAccountInfo>? = null
    private val repository = context.injector.repository()

    fun getMetaData(groupPosition: Int, childPosititon: Int) = getChild(groupPosition, childPosititon).getOrNull()

    override fun getChild(groupPosition: Int, childPosititon: Int) =
        getChildList(groupPosition)!![childPosititon]

    private fun getChildList(groupPosition: Int) = accountMetaDataMap[groupPosition]

    override fun getChildId(groupPosition: Int, childPosition: Int) = childPosition.toLong()

    override fun getChildView(
        groupPosition: Int, childPosition: Int,
        isLastChild: Boolean, convertView: View?, parent: ViewGroup
    ): View {
        val result = convertView ?: layoutInflater.inflate(R.layout.sync_account_row, parent, false)
        val accountMetaData = getChild(groupPosition, childPosition)
        val syncStateView = result.findViewById<ImageView>(R.id.state)
        val labelTextView = result.findViewById<TextView>(R.id.label)
        accountMetaData.onSuccess {
            labelTextView.text = it.toString()
            result.findViewById<View>(R.id.color1)
                .setBackgroundColor(it.color())
            when (getSyncState(groupPosition, childPosition)) {
                SyncState.UNKNOWN, SyncState.ERROR -> syncStateView.visibility = View.GONE
                SyncState.SYNCED_TO_THIS -> {
                    syncStateView.visibility = View.VISIBLE
                    syncStateView.setImageResource(R.drawable.ic_sync)
                }
                SyncState.UNSYNCED, SyncState.SYNCED_TO_OTHER -> {
                    syncStateView.visibility = View.VISIBLE
                    syncStateView.setImageResource(R.drawable.ic_action_sync_unlink)
                }
                SyncState.SEALED -> {
                    syncStateView.visibility = View.VISIBLE
                    syncStateView.setImageResource(R.drawable.ic_lock)
                }
            }
        }.onFailure {
            syncStateView.visibility = View.GONE
            labelTextView.text = it.message
        }
        return result
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        val childList = getChildList(groupPosition)
        return childList?.size ?: 0
    }

    override fun getGroup(groupPosition: Int): Pair<String, Boolean> {
        return syncAccounts[groupPosition]
    }

    fun getBackendLabel(groupPosition: Int): String {
        return getGroup(groupPosition).first
    }

    fun isEncrypted(packedPosition: Long): Boolean {
        val groupPosition = ExpandableListView.getPackedPositionGroup(packedPosition)
        return getGroup(groupPosition).second
    }

    override fun getGroupCount(): Int {
        return syncAccounts.size
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getGroupView(
        groupPosition: Int, isExpanded: Boolean,
        convertView: View?, parent: ViewGroup
    ): View {
        val result = convertView ?: layoutInflater.inflate(R.layout.sync_backend_row, parent, false)
        val group = getGroup(groupPosition)
        val headerTitle = group.first
        result.findViewById<TextView>(R.id.label).text = headerTitle
        result.findViewById<View>(R.id.state).visibility =
            if (group.second) View.VISIBLE else View.GONE
        return result
    }

    override fun hasStableIds(): Boolean {
        return false
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return true
    }

    fun setAccountList(accountList: List<Pair<String, Boolean>>) {
        syncAccounts = accountList
        accountMetaDataMap.clear()
        notifyDataSetChanged()
    }

    fun setAccountMetadata(
        groupPosition: Int,
        accountMetaDataList: List<Result<AccountMetaData>>?
    ) {
        accountMetaDataMap.put(groupPosition, accountMetaDataList)
        notifyDataSetChanged()
    }

    fun hasAccountMetadata(groupPosition: Int): Boolean {
        return accountMetaDataMap[groupPosition] != null
    }

    fun setLocalAccountInfo(localAccountInfo: List<LocalAccountInfo>) {
        this.localAccountInfo = localAccountInfo
        notifyDataSetChanged()
    }

    fun getSyncState(packedPosition: Long): SyncState {
        return getSyncState(
            ExpandableListView.getPackedPositionGroup(packedPosition),
            ExpandableListView.getPackedPositionChild(packedPosition)
        )
    }

    private fun getSyncState(groupPosition: Int, childPosition: Int) =
        getChild(groupPosition, childPosition).map { accountMetaData ->
            localAccountInfo?.find { it.uuid == accountMetaData.uuid() }?.let {
                if (it.isSealed) SyncState.SEALED else when (it.syncAccountName) {
                    getBackendLabel(groupPosition) -> SyncState.SYNCED_TO_THIS
                    null -> SyncState.UNSYNCED
                    else -> SyncState.SYNCED_TO_OTHER
                }
            } ?: SyncState.UNKNOWN
        }.getOrElse { SyncState.ERROR }

    fun getAccountForSync(packedPosition: Long): Account? {
        val groupPosition = ExpandableListView.getPackedPositionGroup(packedPosition)
        val result = getChild(
            groupPosition,
            ExpandableListView.getPackedPositionChild(packedPosition)
        )
        result.onSuccess { accountMetaData ->
            return accountMetaData.toAccount(homeCurrency, getBackendLabel(groupPosition))
        }
        return null
    }

    fun getSyncAccountName(packedPosition: Long): String {
        val groupPosition = ExpandableListView.getPackedPositionGroup(packedPosition)
        return getBackendLabel(groupPosition)
    }

}